### ISSUE TITLE
Make DEIntegrator have type parameters

### ISSUE THREAD

**Issue by Datseris** [JuliaDiffEq/DiffEqBase.jl#181]
(and make its subtypes also obey that)

This will allow dispatching on e.g. the state type of the integrator, which I used a lot in DynamicalSystems. I have opened some PRs to do this, but they will clearly be modified since the plan of how to do it will be suggested by Chris here: @ChrisRackauckas .

**Comment by ChrisRackauckas** [comment]
It should probably be `{alg,inplace,uType,tType}` since those are the critical types that define the compilation process for the differential equation. All differential equations have these quantities, so it should be on the `DEIntegrator` level. 

The PRs have quite a few issues but that's easily fixable. The only real issue I see is that the release process might be difficult. The reason is:

```julia
abstract type MyType{X} end
struct MySubtype <: MyType end
```

that errors. So if DiffEqBase updates, everything needs to update or DifferentialEquations.jl will error. So everything needs and upper bound and a PR in place in order to merge this. This isn't breaking on the user-interface level so it doesn't need to wait for DiffEq 6.0, but it definitely will be an intense release process so it cannot happen before my wedding.

We should get other opinions in here on what types should be in the abstract type. @YingboMa @devmotion ? I think those 4 are the only ones which are common and commonly used for making decisions. We might be able to clean up a good amount of internal code with this. What about the ordering? The most useful ones should go first.

**Comment by Datseris** [comment]
Yeah I'll make new PRs/commits once the plan is set.

`{alg,inplace,uType}` are by far the most useful ones and the ones I have used (actually in DynamicalSystems we were not using `alg` so far). Honestly, is there any direct dispatch use for `tType`?

**Comment by ChrisRackauckas** [comment]
Yes, we use it in PuMaS when autodiffing parts that change event timings. But yes, tType is probably the least used for dispatch and should go last.

**Comment by devmotion** [comment]
Also to me it seems `{alg,inplace,uType,tType}` are the most important types of an integrator. Since we do not want to dispatch on `pType` (as far as I know?) I guess it does not make sense to add it as a type parameter?

>  We might be able to clean up a good amount of internal code with this.

Do you have anything particular in mind?

**Comment by ChrisRackauckas** [comment]
Some of the integrator function implementations do switching based on inplaceness and the algorithm. We can probably just use the part from this signature there.

**Comment by Datseris** [comment]
So is this settled? We got with `{alg,inplace,uType,tType}`?

**Comment by devmotion** [comment]
> Some of the integrator function implementations do switching based on inplaceness and the algorithm. We can probably just use the part from this signature there.

I was curious since e.g. for `ODEIntegrator` in most cases we have to dispatch on the cache if we want to support composite algorithms, so I wasn't sure to which extent we would benefit from the type information.

**Comment by Datseris** [comment]
I will assume that we have agreed to go with `{alg, inplace, uType, tType}` and go ahead with this.  From Slack, Chris said that
> OrdinaryDiffEq, Sundials, StochasticDiffEq, DelayDiffEq, ODEInterfaceDiffEq, DASKR, LSODA, GeometricIntegratorsDiffEq 

should change. I'll start withe diffeq and ode ones.

**Comment by Datseris** [comment]
Btw I'll use `{Alg, IIP, U, T}` to keep things sort.

**Comment by Datseris** [comment]
I'll wait a bit more until Chris has some time to reply here before I move on working more on this. I want to be sure that this is the plan we all agree to move forward with.

**Comment by ChrisRackauckas** [comment]
Yeah, that plan looks good. Though lets wait until after Christmas for doing it since this upgrade will be painful.

**Comment by Datseris** [comment]
@ChrisRackauckas I just cloned the GeometricIntegratorsDiffEq to work on it, but it doesn't seem to use the DiffEq integrators? It doesn't have an `init` call for example.

**Comment by ChrisRackauckas** [comment]
Oh yes, both of those are not setup with the integrator interface.

**Comment by Datseris** [comment]
Added a PR for Delay, only stochastic left. I haven't touched it yet only because I can see that there are more types that what I was expecting:

```
SDEIntegrator{T1,uType,uEltype,tType,P,eigenType,tTypeNoUnits,uEltypeNoUnits,randType,rateType,solType,cacheType,F4,F5,OType,noiseType,EventErrorType} 
```

so maybe Chris wants to say something about those? Also, I don't see an Algorithm type here at all; is it the `cacheType` ?

**Comment by devmotion** [comment]
You can see in https://github.com/JuliaDiffEq/StochasticDiffEq.jl/blob/master/src/integrators/type.jl#L26 that the algorithm type is `T1`.
